testsuite/classlibrary/TestCondVar.sc

```diff
@@ -0,0 +1,386 @@
+		var flags = false!2;
+
+		2.do { |i|
+			fork {
+				c.wait {
+					counters[i] = counters[i] + 1;
+					flags[i]
+				};
+				doneFlags[i] = true;
+			}
+		};
+
+		this.momentarilyYieldUntil { counters == [1, 1] };
+		this.assertEquals(doneFlags, [false, false], "Done flags before first signalAll");
+
+		c.signalAll;
+		this.momentarilyYieldUntil { counters == [2, 2] };
+		this.assertEquals(doneFlags, [false, false], "Done flags before flags[0] set");
+
+		flags[0] = true;
+		c.signalAll;
+		this.momentarilyYieldUntil { counters == [3, 3] };
+		this.assertEquals(doneFlags, [true, false], "Done flags after flags[0] set");
+
+		flags[1] = true;
+		c.signalAll;
+		this.momentarilyYieldUntil { counters == [3, 4] }; // thread 1 already exited
+		this.assertEquals(doneFlags, [true, true], "Done flags after flags[1] set");
+	}
+
+	// waitFor - preconditions for timeoutBeats
+	test_waitFor_timeoutBeats_badValuesThrowErrors {
+		var c = CondVar();
+
+		this.assertException({ c.waitFor(nil) }, Error, "nil");
+		this.assertException({ c.waitFor(TestCondVarBadIntConversion()) }, Error, "Bad int conversion");
+		this.assertException({ c.waitFor(TestCondVarBadFloatConversion()) }, Error, "Bad float conversion");
+		this.assertException({ c.waitFor(inf) }, Error, "inf");
+		this.assertException({ c.waitFor(0/0) }, Error, "NaN");
+	}
+
+	test_waitFor_timeoutBeats_immediateTimeout {
+		var c = CondVar();
+		// Makes use of the fact that the condition must be evaluated at least twice unless we timeout immediately
+		var checkValue = { |val|
+			var counter = 0;
+			var result = c.waitFor(val) { counter = counter + 1; false };
+			this.assertEquals(counter, 1, "waitFor(%): predicate only evaluated once".format(val));
+			this.assert(result.not, "waitFor(%): result".format(val));
+		};
+
+		checkValue.(0);
+		checkValue.(0.0);
+		checkValue.(-1);
+		checkValue.(-inf);
+
+		// Check good conversions too, these convert to 0 and 0.0 respectively
+		checkValue.(TestCondVarGoodIntConversion());
+		checkValue.(TestCondVarGoodFloatConversion());
+	}
+
+	// waitFor - without predicate
+	test_waitFor_withoutPredicate_signalled {
+		var c = CondVar();
+		fork { c.signalOne };
+		// use assertEquals in case waitFor returns a non-boolean
+		this.assertEquals(c.waitFor(1), true);
+	}
+
+	test_waitFor_withoutPredicate_timedOut {
+		var c = CondVar();
+		// use assertEquals in case waitFor returns a non-boolean
+		this.assertEquals(c.waitFor(this.smallTimeSlice), false);
+	}
+
+	// waitFor - with predicate
+	test_waitFor_withPredicate_immediatelyTrue {
+		var c = CondVar();
+		this.assertEquals(c.waitFor(this.smallTimeSlice) { true }, true);
+	}
+
+	test_waitFor_withPredicate_signaledButFalseSoTimedOut {
+		var c = CondVar();
+		var counter = 0;
+		var waitResult;
+		fork { c.signalOne };
+
+		// This wait has to be a bit longer because we need to ensure there's enough time for the spurious wakeup before
+		// timing out.
+		waitResult = c.waitFor(0.1) {
```

Comment:
I've never seen this syntax before. I can understand writing `c.wait { /* */ }` but not so much `c.waitFor(1) { /* */ }`. Maybe I'm just not used to it. What do others think? @joshpar @telephon @jamshark70

Reply:
In a way, we've all seen this syntax before, with a slight difference:

- `if(receiver) { block } { block }`
- `switch(receiver) { block } { block } { block }...`

The difference here is: `receiver.waitFor(arg0) { block }`.

My feeling is that this is not a massive difference, and it doesn't bother me. However, Brian balked at it in the earlier PR (so I'm surprised to see it appear in the test suite).

This is one where, no matter what you do, somebody can complain about something:

- If we recommend `c.wait { xyz }` and `c.waitFor(time, { xyz })`, then the bracketing is inconsistent between the two forms.

- If we recommend `c.wait({ xyz })` and `c.waitFor(time, { xyz })`, someone will be bothered by the double brackets for `wait`.

- If we recommend `c.wait { xyz }` and `c.waitFor(time) { xyz })`... well, the present question is exactly a complaint about this.

My personal opinion is to favor trailing-function syntax in general, especially for methods related to flow of control. (`wait` is not a branching keyword, but CondVar is a control-flow device.) If we are comfortable with `if(x) { y }` then I see no reason to be uncomfortable with `cond.waitFor(num) { y }`. For that matter, I never got in the habit of writing e.g. `do(10) { body }` but I wouldn't object to it (and in some ways I think it reads even better than `10.do`).

TL;DR If it's up to me, I wouldn't change it, but I won't object if it really troubles others.

Reply:
Thanks for the explanation! I agree, we should just keep it as is.